code/game/objects/effects/effect_system/effects_foam.dm

```diff
@@ -122,80 +119,67 @@
 		var/turf/T = get_turf(src)
 		if(isspaceturf(T)) //Block up any exposed space
 			T.PlaceOnTop(/turf/open/floor/plating/foam)
-		for(var/direction in GLOB.cardinals)
-			var/turf/cardinal_turf = get_step(T, direction)
+		for(var/i in 1 to GLOB.cardinals.len)
+			var/turf/cardinal_turf = get_step(T, GLOB.cardinals[i])
 			if(get_area(cardinal_turf) != get_area(T)) //We're at an area boundary, so let's block off this turf!
 				new/obj/structure/foamedmetal(T)
 				break
 	flick("[icon_state]-disolve", src)
 	QDEL_IN(src, 5)
 
 /obj/effect/particle_effect/foam/process()
-	lifetime--
-	if(lifetime < 1)
-		kill_foam()
-		return
-
-	var/fraction = 1/initial(reagent_divisor)
-	for(var/obj/O in range(0,src))
-		if(O.type == src.type)
-			continue
-		if(isturf(O.loc))
-			var/turf/T = O.loc
-			if(T.intact && O.level == 1) //hidden under the floor
-				continue
-		if(lifetime % reagent_divisor)
-			reagents.reaction(O, VAPOR, fraction)
-	var/hit = 0
-	for(var/mob/living/L in range(0,src))
-		hit += foam_mob(L)
-	if(hit)
-		lifetime++ //this is so the decrease from mobs hit and the natural decrease don't cumulate.
-	var/T = get_turf(src)
-	if(lifetime % reagent_divisor)
-		reagents.reaction(T, VAPOR, fraction)
-
-	if(--amount < 0)
-		return
-	spread_foam()
+	var/turf/my_turf = get_turf(src)
+	if(my_turf)
+		var/hit = 0
+		for(var/i in 1 to my_turf.contents.len)
+			var/atom_to_hit = my_turf.contents[i]
+			if(istype(atom_to_hit, /obj))
+				var/obj/O = atom_to_hit
+				if(O.type == src.type)
+					continue
+				if(my_turf.intact && O.level == 1)
+					continue
+				reagents.reaction(O, TOUCH)
+			if(istype(atom_to_hit, /mob/living))
+				var/mob/living/L = atom_to_hit
+				hit += foam_mob(L)
+		reagents.reaction(T, TOUCH)
+		amount--
+		if(amount)
+			spread_foam()
+		
 
 /obj/effect/particle_effect/foam/proc/foam_mob(mob/living/L)
-	if(lifetime<1)
-		return 0
 	if(!istype(L))
 		return 0
-	var/fraction = 1/initial(reagent_divisor)
-	if(lifetime % reagent_divisor)
-		reagents.reaction(L, VAPOR, fraction)
-	lifetime--
+	reagents.reaction(L, TOUCH)
 	return 1
 
 /obj/effect/particle_effect/foam/proc/spread_foam()
 	var/turf/t_loc = get_turf(src)
-	for(var/turf/T in t_loc.GetAtmosAdjacentTurfs())
-		var/obj/effect/particle_effect/foam/foundfoam = locate() in T //Don't spread foam where there's already foam!
-		if(foundfoam)
-			continue
-
-		if(is_type_in_typecache(T, blacklisted_turfs))
-			continue
-
-		for(var/mob/living/L in T)
-			foam_mob(L)
-		var/obj/effect/particle_effect/foam/F = new src.type(T)
-		F.amount = amount
-		reagents.copy_to(F, (reagents.total_volume))
-		F.add_atom_colour(color, FIXED_COLOUR_PRIORITY)
-		F.metal = metal
-
-
-/obj/effect/particle_effect/foam/temperature_expose(datum/gas_mixture/air, exposed_temperature, exposed_volume)
-	if(prob(max(0, exposed_temperature - 475))) //foam dissolves when heated
-		kill_foam()
```

Comment:
This is a behavior change that isn't mentioned in your changelog

Reply:
ty for reminding me